### PR TITLE
Search improve

### PR DESCRIPTION
**Pull request by romanesko**
https://github.com/linagora/Twake/issues/2007

### CODE REVIEW

File: twake/backend/node/src/core/platform/services/search/adapters/elasticsearch/index.ts [hunk 1]
```diff
@@ -252,9 +252,13 @@ export default class ElasticSearch extends SearchAdapter implements SearchAdapte
     };
 
     let esResponse: ApiResponse;
+
     if (options.pagination.page_token) {
       esResponse = await this.client.scroll(
-        { scroll_id: options.pagination.page_token, ...esParamsWithScroll },
+        {
+          scroll_id: options.pagination.page_token,
+          // ...esParamsWithScroll
```
Comment (RomaricMourgues) [inline twake/backend/node/src/core/platform/services/search/adapters/elasticsearch/index.ts:260]: Could you explain to me why is it removed?
And then depending on the response I suppose there is two option here:
- Put it back
- Remove it completely
😛

File: twake/backend/node/src/services/messages/services/views/service.ts [hunk 2]
```diff
@@ -222,6 +222,7 @@ export class ViewsService implements MessageViewsServiceAPI {
         ...(options.companyId ? { $in: [["company_id", [options.companyId]]] } : {}),
         ...(options.workspaceId ? { $in: [["workspace_id", [options.workspaceId]]] } : {}),
         ...(options.channelId ? { $in: [["channel_id", [options.channelId]]] } : {}),
+        // TODO: implement boolean for hasFiles
```
Comment (RomaricMourgues) [inline twake/backend/node/src/services/messages/services/views/service.ts:225]: You can already do it:

```
async search(
    pagination: Pagination,
    options?: SearchMessageOptions,
    context?: ExecutionContext,
  ): Promise<ListResult<Message>> {
    return await this.searchRepository.search(
      {
      //Here:
        has_file: myBoolean
      },
      {
        pagination,
        ...(options.companyId ? { $in: [["company_id", [options.companyId]]] } : {}),
        ...(options.workspaceId ? { $in: [["workspace_id", [options.workspaceId]]] } : {}),
        ...(options.channelId ? { $in: [["channel_id", [options.channelId]]] } : {}),
```

Comment (RomaricMourgues) [reply twake/backend/node/src/services/messages/services/views/service.ts:225]: Same place for the `sender`

File: twake/backend/node/src/services/messages/web/controllers/views.ts [hunk 3]
```diff
@@ -1,26 +1,65 @@
...
+    return this.data[key];
+  }
+
+  set(key: string, value: { [key: string]: boolean }) {
+    this.expiration[key] = SimpleChannelsCache.now() + this.minutesTTL * 1000;
+    this.data[key] = value;
+  }
+}
+
+const simpleChannelsCache = new SimpleChannelsCache(5);
```
Comment (RomaricMourgues) [inline twake/backend/node/src/services/messages/web/controllers/views.ts:54]: Actually we already have a cache available `import NodeCache from "node-cache";` you can see it in action here: twake/backend/node/src/services/user/services/users/service.ts

**Review by RomaricMourgues — CHANGES_REQUESTED** [review]
Nice to have you back on the code!

I have some feedback, that's normal after this delay. There is mainly small improvements asked on my comments.

I also need to discuss with you about the current strategy of getting all the channels first and then looking at this big list on the search results. Take a look at how I did it for listing messages and adding the include_users in it: `includeUsersInDirectChannel` in `src/services/channels/services/channel/service.ts`

Thanks for the awesome work on tests and global structure of the code!

File: twake/backend/node/src/core/platform/services/search/adapters/elasticsearch/index.ts (hunk 1 above)
Comment (romanesko) [reply twake/backend/node/src/core/platform/services/search/adapters/elasticsearch/index.ts:260]: It doesn't work with query in `_search/scroll`. It expects only `scroll_id` that refers to the previous query, and it falls with error if query exists in this request. I tested on a local elastic search and found this pattern. The documentation mentions it in passing, but it doesn't work with `query` anyway. So I suppose we can remove it completely.

File: twake/backend/node/src/services/messages/web/controllers/views.ts (hunk 3 above)
Comment (romanesko) [reply twake/backend/node/src/services/messages/web/controllers/views.ts:54]: 🤦‍♂️ … 
mine is pretty elegant too )
ok, I'll change it to node-cache

File: twake/backend/node/src/services/messages/services/views/service.ts (hunk 2 above)
Comment (romanesko) [reply twake/backend/node/src/services/messages/services/views/service.ts:225]: Couldn't handle with has_files actually, need your help.

**Comment by RomaricMourgues** [comment]
Yes it was pretty elegant unfortunately 😬

**Comment by RomaricMourgues** [comment]
Interesting ! Scroll_id probably contain the information of pagination and stuff already I suppose. Thanks for the information you can remove it then ;)

**Comment by RomaricMourgues** [comment]
If you can share the generated elastic search query which should appear in the logs I can probably help
I mean if the problem is in the elastic search query of course

File: twake/backend/node/src/services/messages/entities/messages.search.ts [hunk 4]
```diff
@@ -24,9 +26,11 @@ export default {
   esMapping: {
     properties: {
       text: { type: "text" },
+      user_id: { type: "text" },
```
Comment (RomaricMourgues) [inline twake/backend/node/src/services/messages/entities/messages.search.ts:29]: If you want to search using this format:
`...(options.sender ? { $in: [["user_id", [options.sender]]] } : {}),`
You need to use this type:
`user_id: { type: "keyword" }`

File: twake/backend/node/src/services/channels/services/member/service.ts [hunk 5]
```diff
@@ -583,17 +618,32 @@ export class Service implements MemberService {
...
-      workspace_id: channel.workspace_id,
-      user_id: user.id,
-    });
+    const get = () =>
+      this.get({
+        channel_id: channel.id,
+        company_id: channel.company_id,
+        workspace_id: channel.workspace_id,
+        user_id: user.id,
+      });
```
Comment (RomaricMourgues) [inline twake/backend/node/src/services/channels/services/member/service.ts:636]: I don't think we still need a function in a function, you can move this code block to the end return (the variable `const get` is only used once

File: twake/backend/node/src/services/messages/web/controllers/views.ts [hunk 6]
```diff
@@ -113,27 +114,73 @@ export class ViewsController {
...
-        context,
-      )
-      .then(a => a.getEntities());
+        50,
+      );
+      if (!isChannelMember) continue;
+
+      // if (request.query.sender && msg.user_id != request.query.sender) continue;
+
+      // if (request.query.has_files && !(msg.files || []).length) continue;
```
Comment (RomaricMourgues) [inline twake/backend/node/src/services/messages/web/controllers/views.ts:171]: This comments should be removed when tests pass I suppose :)

File: twake/backend/node/src/services/messages/entities/messages.search.ts (hunk 4 above)
Comment (romanesko) [reply twake/backend/node/src/services/messages/entities/messages.search.ts:29]: hm… but it works this way…
ok, I'll change

Comment (RomaricMourgues) [reply twake/backend/node/src/services/messages/entities/messages.search.ts:29]: I didn't know it worked haha :p 
But another point for me is that everything you put in "text" will be searchable in the "query" like if the message text is "I love the option b514" and sender is "0ac1d15a-a5dd-11ec-b514-476f788a5663" then searching for the text "b514" will return all the messages from this sender.
So change it to `keyword` anyway ;)

**Comment by RomaricMourgues** [comment]
For code it is ok for me, but tests don't like it yet :)

**Comment by romanesko** [comment]
Is it possible that tests use non-elastic-search configuration? Can't figure out

**Pull request merged**